xapian-letor/ranker/ranker.cc

```diff
@@ -164,7 +164,12 @@ load_relevance(const std::string & qrel_file)
 	    if (*end) {
 		throw LetorParseError("Could not parse relevance label in Qrel file at line:" + str(qrel_count));
 	    }
-	    qrel1[token[0]].insert(make_pair(token[2], label));
+	    // Map the file name in the qrel file, to the doc-id
+	    Xapian::PostingIterator db_iter = (db.postlist_begin("U/" + token[2]));
+	    if (db_iter != db.postlist_end("U/" + token[2])) {
+		int doc_id = (*db_iter);
+		qrel1[token[0]].insert(make_pair(doc_id, label));
+	    }
```

Comment:
If we don't find the document is quietly ignoring it the appropriate response?

Reply:
Should we log that?

Reply:
I feel throwing an error of "File 'X' in qrel not found" would be a decent response, throw an error and ignore the current file name, and proceed to the next. As the current xapian-prepare-training file "expects" the format to be correct, if not, it shows an LetorParseError logged as:
`"Could not parse relevance label in Qrel file at line:"` after which the code exits.
But that is only when the format of the Qrel file is wrong, at cases where the input file is "not in the wrong format" it ends up writing nothing to the training-file and an empty training file is created. I think a sort-of log line or something should be helpful. Without logging anything, users would have to start digging in, where the error might have come from.

Reply:
> throw an error and ignore the current file name, and proceed to the next

Once you've thrown an exception, you can't just proceed to the next entry because throwing an exception jumps to wherever the most recent enclosing `try`/`catch` which matches is, and then execution continues from there (or the program terminates if there isn't a `try`/`catch` which matches).

Your reasoning does suggest an exception makes most sense though.

Thinking this through, the approach we have here of parsing a file in a particular format to get the relevance judgements seems a little awkward - there's not particular reason why you have to store the judgements in a file even (could be in a database, or calculated on demand - currently such setups would be forced to generate a temporary file of the required format).

As we've found here, it means that error reporting is less flexible.

Also, the code to map the letor "docid" (which is the filename in this case, but that's probably not universally true) to a `Xapian::docid` is rather application-specific.  Most Xapian databases don't have the `U`-prefix terms which this code assumes.

Ultimately I wonder if a better API here would be one which allows the caller to create a "relevance judgements" object and then add judgements to it.  The code to parse and load a "qrel format" file would then go into a tool which iterates over the qrel file and for each line of data calls this new API to add a judgement.  The application-specific code to determine the `Xapian::docid` goes in the application, and the application can log and skip lines that don't have matching documents.

So I think probably throw an exception here for now, and at some point this API can get replaced with a cleaner one.